File: R/cantilever_emulator.R
# Set up R

library(data.table)
library(rstan)
library(maximin)
library(matrixStats)
library(plot3D)
library(plot3Drgl)

# Set current working directory. This should be modified to match the directory
# at which the stan code and any data is stored
setwd("C:/Users/cs2361/Documents/Bayes_Cantilever_Calibration/R/")

# include functions which are called in this code
source("src/cantileverBeam.R")

#-------------------------------------------------------------------------------

# Setup model parameters

# First we set up the parameters
n_simulations = 30   # number of simulations
N_plot = 100         # Number of points at which beam deflection is plotted
N_sim_data = 30      # number of output points in each simulation
n_maximin = 30       # Number of model output data points which are to be retained for training the emulator. I use a smaller number here as the code runs very slowly if I use everything

P <- 10 # Applied load (N)
L <- 1 # length
E <- 68E9 # Young's modulus (Aluminium)
d <- 0.01 # depth
b <- 0.04 # breadth

#-------------------------------------------------------------------------------

# Set up simulations. 

# Sample across the 1D depth variable by splitting its range into equal intervals,
# then sampling at random within each interval.

# define n_simulations intervals of equal size in range [-0.1,0.1]
int_start = (0:(n_simulations-1))/n_simulations*0.2-0.1
int_end = (1:n_simulations)/n_simulations*0.2-0.1
# generate n_simulations uniform random samples
int_rnd = runif(n = n_simulations, min = 0, max = 1)
# place random points within their intervals. These are the values of d for the emulator training data
d_simulation = data.table(d*((int_end-int_start)*int_rnd + int_start))
colnames(d_simulation) = "d"

# Define the points at which the simulation output is generated. Here we take N_sim_data equally spaced points
# between x = 0 and x = L
x_sim = seq(0, L, length.out = N_sim_data)

# We have a total of n_simulations x N_sim_data data points. Arrange these into
# a table
# replicate the simulation coordinates for each n_simulations simulation
dt_sim_coords = data.table(rep(x_sim,n_simulations))
colnames(dt_sim_coords) = "x_sim"
# initialise matrix for storing n_simulations x N_sim_data output points
dt_all_simulation = matrix(0, ncol = n_simulations, nrow = n_simulations)
# run the cantilever model n_simulations times
for (i in 1:n_simulations){
  dt_all_simulation[,i] = beamDeflection(x_sim,E,b,d + as.numeric(d_simulation[i]),P,L)
}
# reshape all simulated displacements into a vector, rather than a 2d grid. These are stored as follows:
#[x_1-sim_1,...,x_N_sim_data-sim_1, x_1-sim_2, ..., x_N_sim_data-sim_N_simulations]
dt_all_simulation = as.vector(dt_all_simulation)
dt_all_simulation = data.table(dt_all_simulation)
colnames(dt_all_simulation) = "D"
# Also duplicate entries in the DoE for the depth to match this stored output
# store in format: [Sim1 x N_sim_data, Sim2 x N_sim_data, ... , sim_N_simulations x N_N_sim_data]
idx = rep(1:n_simulations, rep(N_sim_data,n_simulations))
d_simulation = d_simulation[idx, ]

# Combine all input data into one table
XT_sim = cbind(dt_sim_coords, d_simulation)

# For speed cherry-pick data points using the maximin criterion in the x-d space
# transform inputs onto unit hypercube prior to the maximin design. 
minx = as.matrix(colMins(as.matrix(XT_sim)))
minx = matrix(rep(minx,n_simulations*N_sim_data),ncol=nrow(minx),byrow=TRUE)
maxx = as.matrix(colMaxs(as.matrix(XT_sim)))
maxx = matrix(rep(maxx,n_simulations*N_sim_data),ncol=nrow(maxx),byrow=TRUE)
XT_trans = (XT_sim - minx)/(maxx-minx)

# Use the maximin library to determine an index of the inputs which maximise the maximin criterion
sim_ind = sort(maximin.cand(n = n_maximin, Tmax = 100*n_maximin, Xcand = as.matrix(XT_trans))$inds)
# Retain only the data points with the above index
dt_simulation = cbind(XT_sim[sim_ind,], dt_all_simulation[sim_ind])
XT_sim = XT_sim[sim_ind,]

# If desired, use the below code to write training data to csv files
# write.csv(XT_sim, "benchmark_inputs.csv", row.names = FALSE)
# write.csv(dt_simulation, "benchmark_outputs.csv", row.names = FALSE)

# If desired to reproduce past results, use the below code to read benchmark data from csv file
# XT_sim = read.table("benchmark_inputs.csv", header = TRUE, sep=',')
# dt_simulation = read.table("benchmark_outputs.csv", header = TRUE, sep=',')

#-----------------------------------------------------------------------------

# plot Design of Experiments used to generate training data
dev.new(noRStudioGD = TRUE) # generate plot in new window
plot(XT_sim$x_sim,XT_sim$d,"p",main="Design of Experiments", 
     xlab = "x", ylab = "thickness deviation", xlim=c(0,L), 
     ylim=c(-0.1*d,0.1*d), pch=4, col="blue")

# plot legend
legend(x = "topright",legend = "training data point",col="blue",pch=4)

#-----------------------------------------------------------------------------

# Generate a grid of points for making emulator predictions combining a regular 
# grid, and the training data.

# Generate a sequence of points across x and d.
x_grid <- seq(0, L, length.out = 10)
t_grid <- seq(-0.1*d, 0.1*d, length.out = 10)

# Add in unique values of x and d from design of experiments to these points
x_grid = unique(c(x_grid,XT_sim$x_sim))
t_grid = unique(c(t_grid,XT_sim$d))
# Sort points in ascending order
x_grid = x_grid[order(x_grid)]
t_grid = t_grid[order(t_grid)]
# Expand points into a grid
xt <- expand.grid(x = x_grid, t = t_grid)

#-----------------------------------------------------------------------------

# Put the data into the correct format for input to stan, and run stan.

# Set up the environment for the Stan model to run in parallel. Taken from:
# https://betanalpha.github.io/assets/case_studies/gaussian_processes.html#21_Simulating_From_A_Gaussian_Process
rstan_options(auto_write = TRUE)
options(mc.cores = parallel::detectCores())
parallel:::setDefaultClusterOptions(setup_strategy = "sequential")
util = new.env()
par(family="CMU Serif", las=1, bty="l", cex.axis=1, cex.lab=1, cex.main=1,
    xaxs="i", yaxs="i", mar = c(5, 5, 3, 5))

# Note, the below code is modified from the implementation at:
# https://github.com/adChong/bc-stan/blob/master/src/main.R

# get dimensions of the various datasets
p = 1                   # number of "controlled" inputs, x
q = 1                   # number of "uncontrolled" inputs, t
m = nrow(XT_sim)        # sample size of computer simulation data

# extract data from various tables
eta = dt_simulation$D           # simulation output (emulator training data)
xc = as.matrix(XT_sim$x_sim)    # "controlled" input
tc = as.matrix(XT_sim$d)        # "uncontrolled" input
x_pred = as.matrix(xt$x)  # "controlled" input
t_pred = as.matrix(xt$t)  # "uncontrolled" input
n_pred = nrow(x_pred)   # store required number of predictions

# standardise ouputs to have zero sample mean unit sample variance. 
eta_mu = mean(as.matrix(eta), na.rm = TRUE) # mean value
eta_sd = sd(as.matrix(eta), na.rm = TRUE)   # standard deviation
eta = (eta - eta_mu)/eta_sd

# Standardise controlled and uncontrolled inputs to lie on [0,1]
x_min = as.matrix(colMins(xc))
x_max = as.matrix(colMaxs(xc))
xc = (xc - matrix(rep(x_min,m,ncol=p,byrow=TRUE)))/matrix(rep(x_max-x_min,m,ncol=p,byrow=TRUE))
x_pred = (x_pred - matrix(rep(x_min,n_pred,ncol=p,byrow=TRUE)))/matrix(rep(x_max-x_min,n_pred,ncol=p,byrow=TRUE))
t_min = as.matrix(colMins(tc))
t_max = as.matrix(colMaxs(tc))
tc = (tc - matrix(rep(t_min,m,ncol=q,byrow=TRUE)))/matrix(rep(t_max-t_min,m,ncol=q,byrow=TRUE))
t_pred = (t_pred - matrix(rep(t_min,n_pred,ncol=q,byrow=TRUE)))/matrix(rep(t_max-t_min,n_pred,ncol=q,byrow=TRUE))

# create data as list for input to Stan
stan_data = list(m=m, n_pred=n_pred, p=p, q=q, eta=c(as.matrix(eta)), xc=as.matrix(xc), 
                 x_pred=as.matrix(x_pred), tc=as.matrix(tc), t_pred=as.matrix(t_pred))

# run code for fitting Bayesian emulator, and making subsequent predictions
fit = stan(file = "stan/Bayesian_Emulator.stan",
            data = stan_data,
            iter = 4000,
            chains = 3)

#---------------------------------------------------------------------------------------------------------------------

# This section of code for post-processing, and generation of plots of predictions

# plot traceplots, excluding warm-up
dev.new(noRStudioGD = TRUE)  # generate plots in separate window
stan_trace(fit, pars = c("rho_eta", "lambda_eta"))

# summarise results to check convergence
print(fit, pars = c("beta_eta", "lambda_eta","f_pred"))

# Extract data from STAN model
samples <- extract(fit)
N_samples = dim(samples$rho_eta)[1] # get total number of samples

# extract predicted displacements and transform these onto their correct scale
# (recalling that observations were transformed to have zero sample mean, and 
# unit sample variance)
f_pred = samples$f_pred
f_trans = f_pred*eta_sd + eta_mu

# Produce plots of posterior and prior distributions of correlation lengths,
# (rho), and precision (lambda)
dev.new(noRStudioGD = TRUE)
par(mfrow = c(1,3))
axis_labels = list("rho_eta1","rho_eta2")
titles = list("correlation length x coordinate","correlation length depth")
rho_plot = seq(0,1, length.out = 100)
for (i in 1:2){
  # Plot histogram of posterior distribution
  hist(samples$rho_eta[,i],
       main = titles[i],
       xlab = axis_labels[i],
       col = "firebrick1",
       breaks = 25,
       freq = FALSE,
       xlim = c(0,1), 
       cex.lab=1.5,
       cex.axis=1.5)
  # Plot prior distribution
  prior_plot = dbeta(rho_plot,shape1=1,shape2=0.3)
  lines(rho_plot,prior_plot,lwd=3,col="blue")
}

# marginal precision of emulator
# plot posterior
hist(samples$lambda_eta,
     main = "lambda_eta",
     xlab = "lambda_eta",
     col = "firebrick1",
     breaks = 25,
     freq = FALSE,
     xlim = c(0,2.5),
     cex.axis=1.5,
     cex.lab=1.5)
# plot prior
lambda_plot = seq(0,2.5, length.out = 100)
prior_plot = dgamma(lambda_plot,shape=10,rate=10)
lines(lambda_plot,prior_plot,lwd=3,col="blue")

# plot an individual sample prediction
samInd = 1; # index of which sample to plot (modify to look at samples)
# Create input and output grids for producing surface plots
x_surf <- matrix(xt$x, nrow = length(x_grid), ncol = length(t_grid)) # x coordinate
y_surf<- matrix(xt$t, nrow = length(x_grid), ncol = length(t_grid)) # beam depth
z_surf <- matrix(f_trans[samInd,], nrow = length(x_grid), ncol = length(t_grid)) # emulator sample

dev.new(noRStudioGD = TRUE) # generate plot in a new window
scatter3D(dt_simulation$x_sim, dt_simulation$d, dt_simulation$D, main = "Individual Sample Realisation", xlab = "x (m))", xlim = c(0, L), ylab = "thickness deviation", ylim = c(-0.1*d,0.1*d), zlab = "displacement", zlim = c(min(z_surf),max(z_surf)), col = "black", add = FALSE, clab = "displacement",pch=18,bty="g",ticktype = "detailed",nticks = 6,surf = list(x = x_surf, y = y_surf, z = z_surf, facets = TRUE, col = ramp.col(c("yellow","red"))))
# also create an interactive object which may be rotated etc
plotrgl()

# Plot the mean prediction to averages across all of the uncertainty. 
f_trans_mu = colMeans(f_trans)
# create output grid for surface plot
z_surf <- matrix(f_trans_mu, nrow = length(x_grid), ncol = length(t_grid))
dev.new(noRStudioGD = TRUE) # generate plot in a new window
scatter3D(dt_simulation$x_sim, (d+dt_simulation$d)*1000, dt_simulation$D, main = "Emulator Mean", xlab = "x (m)", xlim = c(0, L), ylab = "depth (mm)", ylim = c(900*d,1100*d), zlab = "displacement", zlim = c(min(z_surf),max(z_surf)), col = "black", add = FALSE, clab = "displacement",pch=18,bty="g",ticktype = "detailed",nticks = 6,surf = list(x = x_surf, y = 1000*(d+y_surf), z = z_surf, facets = TRUE, col = ramp.col(c("yellow","red"))),cex.lab=1.5,cex.axis=1.5)
plotrgl()

# plot standard deviations of samples.
f_trans_sd = colSds(f_trans) # determine the standard deviation at each point
# create output grid for surface plot
z_surf <- matrix(f_trans_sd, nrow = length(x_grid), ncol = length(t_grid))
dev.new(noRStudioGD = TRUE) # generate plot in a new window
scatter3D(dt_simulation$x_sim, dt_simulation$d, rep(0,n_maximin), main = "Emulator Uncertainty", xlab = "x (m))", xlim = c(0, L), ylab = "thickness deviation", ylim = c(-0.1*d,0.1*d), zlab = "displacement", zlim = c(0,max(z_surf)), col = "black", add = FALSE, clab = "displacement",pch=18,bty="g",ticktype = "detailed",nticks = 6,surf = list(x = x_surf, y = y_surf, z = z_surf, facets = TRUE, col = ramp.col(c("yellow","red"))))
plotrgl()

File: python/cantilever_emulator.py
# Fits a Gaussian process emulator to the point-wise displacement of a simple cantilever beam
# with uncertain depth

import sys
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib import rcParams
import pymc as pm
import arviz as az
from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.gaussian_process.kernels import ConstantKernel, RBF
from scipy.stats import halfnorm, invgamma

src_path = "src/"
sys.path.insert(0, src_path)

from cantilever_beam import *  # Import cantilever model
from LHS_Design import transformed_LHS  # Import Latin Hypercube module
from maximin import *
from transform_input_output import normalise_inputs, standardise_output, rescale_output
import utils

if __name__ == "__main__":

    # ------------------------------------------------------------------------------
    # Define space of inputs, generate training samples, and set up model parameters
    # ------------------------------------------------------------------------------

    E = 68E9 # Young's modulus
    P = 10.0 # Applied load
    b = 0.04 # Beam width
    L = 1.0  # Beam length
    d = 0.01 # Nominal value of  beam height d

    N_train = 25  # Number of required training data points
    
    # Define uncertain input prior values
    # Would be better in Pandas.
    inputs = [
        ['d', 'Uniform', 0.9*d, 1.1*d, True],
    ]
    x_train = transformed_LHS(inputs, N_train, sampler_package="scikit-optimize", sampler_kwargs={"lhs_type":"classic","criterion":"maximin", "iterations":100})
    inp_str = [inp[0] for inp in inputs] # List of input variable names
    inp_str.append("x")

    N_plot = 100   # Number of points at which beam deflection is plotted
    N_maximin = 50 # Number of model output data points which are to be retained for training the emulator 
    # Generate GP prediction sample, not just mean and standard deviation (this is 
    # expensive if predictions are needed at a lot of points)
    gp_sample_pred = False 
    
    # Set plotting parameters
    utils.set_plot_params()

    # ------------------------------------------------------------------------------
    #           Run the cantilever beam model for Design of experiments
    # ------------------------------------------------------------------------------
    
    # Define a list of coordinates at which the simulation output is generated.
    sim_coords = [i/(N_plot-1)*L for i in range(N_plot)]
    
    # Run the cantilever model n_simulations times
    y_all = np.empty(shape = (N_plot,N_train))
    for i, x_i in enumerate(x_train):
        y_all[:,i] = cantilever_beam(x=sim_coords,E=E,b=b,d = x_i[0],P=P, L=L)

    # Plot the output
    fig, ax = plt.subplots()
    ax.plot(sim_coords,y_all)
    ax.set_xlabel("x (m)")
    ax.set_ylabel("Displacement (m)")
    ax.set_title("Beam displacement across training samples")

    # Reshape all n_simulations x N_train points into a vector
    y_all = y_all.T.reshape((-1))
    
    # Create a matrix matching each training data point (x-coordinate and depth) to each output value
    sim_coords = np.array(sim_coords,ndmin=2).T
    x_all = np.concatenate((np.repeat(x_train,N_plot,axis=0), np.tile(sim_coords,(N_train,1))), axis=1)

    # For speed we need to cherry-pick points. Choose observations to satisfy the maximin criterion.
    #  Currently not very efficient - could speed things up by optimising choice of replacement. See R maximin documentation
    # First transform inputs onto unit hypercube (note - we can also use scikit-learn.preprocessing)
    x_all_trans, *_ = normalise_inputs(x_all)
    x_maximin, ind, d_min = maximin(x_all_trans, N_maximin, 15000, replace_method="random")
    # Extract untransformed inputs and outputs for the maximin sample
    x_train = x_all[ind,:]
    y_train = y_all[ind]
    
    x_train = pd.DataFrame(x_train, columns = inp_str)
    y_train = pd.Series(y_train, name="Displacement")

# ------------------------------------------------------------------------------
#                        Plot the Design of Experiments
# ------------------------------------------------------------------------------
    
    N_grades = 8 # Number of different values to divide the output into
    plot_frame = pd.concat((x_train, y_train),axis=1)
    # Bin the displacement into intervals, and create a new categorical column
    # in the dataframe stating which interval each point lies within
    plot_frame["Category"] = pd.cut(plot_frame["Displacement"],N_grades)
    # Create a pairs plot of the training data, coloured according to the displacement of each point
    fig2, axes2 = plt.subplots(1,2, sharey = True)
    fig2.suptitle("Training data before and after maximin search")
    sns.scatterplot(x=inp_str[0], y=inp_str[1], data=plot_frame, ax=axes2[1], hue="Category", legend=False, palette = sns.color_palette("viridis",N_grades))
    # There are only two inputs so it is more appropriate to use scatterplot thann pairplot
    plot_frame = pd.concat((pd.DataFrame(x_all, columns = inp_str),pd.DataFrame(y_all,columns=["Displacement"])),axis=1)
    plot_frame["Category"] = pd.cut(plot_frame["Displacement"],N_grades)
    sns.scatterplot(x=inp_str[0], y=inp_str[1], data=plot_frame, ax=axes2[0], hue="Category", legend=False, palette = sns.color_palette("viridis",N_grades))
    
# ------------------------------------------------------------------------------
#          Generate a set of points at which predictions are required
# ------------------------------------------------------------------------------

    # Generate a sequence of points across x and t
    n_grid = 15
    x_grid = np.linspace(min(sim_coords),max(sim_coords), num = n_grid).reshape(-1)
    t_grid = np.linspace(0.9*d,1.1*d, num = n_grid)

    # Take the union of these points with the training data to verify that uncertainty is zero at these points
    x_grid = np.sort(np.unique(np.concatenate((x_grid,x_train["x"].to_numpy()))))
    t_grid = np.sort(np.unique(np.concatenate((t_grid,x_train["d"].to_numpy()))))
    # Expand points into a grid
    x_grid, t_grid = np.meshgrid(x_grid,t_grid)
    
    #x_pred = np.concatenate((t_grid.reshape(-1,1),x_grid.reshape(-1,1)),axis=1)
    x_pred = pd.DataFrame({"d" : t_grid.reshape(-1), "x" : x_grid.reshape(-1)})
    N_pred = x_pred.shape[0]

# ------------------------------------------------------------------------------
#                           Standardise the data
# ------------------------------------------------------------------------------

    # Standardise outputs to have zero mean and unit sample variance
    y_trans, y_mu, y_sd = standardise_output(y_train)

    # Normalise inputs such that training data is on the unit hypercube
    x_trans, x_min, x_max = normalise_inputs(x_train)

    # Normalise test data in the same way as the training data for consistency
    x_pred_trans = normalise_inputs(x_pred, x_min = x_min, x_max = x_max)
    
#-------------------------------------------------------------------------------
#              Fit emulator using Maximum Likelihood Estimation
#-------------------------------------------------------------------------------

    print("\nEstimating hyperparameters via MLE using scikit-learn\n")

    # Determine emulator hyperparameters which MLE versions from scikit-learn
    # Construct a Kernel: R(X,X') = sigma^2*exp(-(X-X')/(2*LS^2)
    kernel = ConstantKernel(constant_value=1.0,constant_value_bounds=(1e-5,100.0))*RBF(length_scale=tuple(0.5 for i in range(2)), length_scale_bounds=(1e-5,10.00))
    # Create a Gaussian Process regressor object and fit to training data
    gp_MLE = GaussianProcessRegressor(kernel=kernel, alpha = 1e-8, normalize_y = False, n_restarts_optimizer = 50).fit(x_trans, y_trans)
    # Return predictive mean and standard deviation
    mu_pred_MLE, sigma_pred_MLE = gp_MLE.predict(x_pred_trans, return_std=True)
    # Rescale predictions
    mu_pred_MLE = rescale_output(mu_pred_MLE, mu_y = y_mu, sigma_y = y_sd)
    sigma_pred_MLE = rescale_output(sigma_pred_MLE, sigma_y=y_sd, std=True)
    # Extract hyperparameters of fitted Gaussian process
    sigma_em_MLE = np.sqrt(gp_MLE.kernel_.get_params()["k1__constant_value"])
    ls_MLE = gp_MLE.kernel_.get_params()["k2__length_scale"]

#-------------------------------------------------------------------------------
#                    Fit emulator using Bayesian inference 
#-------------------------------------------------------------------------------

    print("\nEstimating hyperparameters via Bayesian inference\n")
    # Create pymc model
    with pm.Model() as emulator_model:
    
        # Priors on emulator hyperparameters and noise parameter
        # Updata to improve convergence. Stan guidance on GP priors:
        #'https://github.com/stan-dev/stan/wiki/Prior-Choice-Recommendations#priors-for-gaussian-processes'
        # Example I'm following:
        # https://www.pymc.io/projects/examples/en/latest/gaussian_processes/GP-Latent.html
              
        # Scale parameter
        sigma_em = pm.HalfNormal("sigma_em", sigma = 1.0)
                
        # Define correlation length parameters
        ls = pm.InverseGamma("ls", alpha = 4.0, beta = 4.0, shape = 2)
        # Define mean and covariance functions
        mean_func = pm.gp.mean.Zero()
        cov_func = pm.gp.cov.Constant(sigma_em**2) * pm.gp.cov.ExpQuad(2,ls=ls)
       
        # Implementation using multivariate normal:
        # y_ = pm.MvNormal("y", mu=mean_func(x_trans).eval(), cov=cov_func(x_trans).eval(), observed = y_trans)
                        
        # Alternative example using marginal likelihood GP (more stable)
        gp = pm.gp.Marginal(mean_func = mean_func, cov_func = cov_func)
        y_ = gp.marginal_likelihood("y", X = x_trans.values, y = y_trans.values, sigma = 1e-8)
    
        # Draw samples from posterior
        idata = pm.sample(3000, target_accept=0.9)

    # Produce trace plots    
    az.plot_trace(idata, combined=True, figsize=(10, 8))


#------------------------------------------------------------------------------
#                 Plot histograms of the posterior marginals
#------------------------------------------------------------------------------

    # Extract posterior samples and plot histograms
    ls = idata.posterior["ls"].values.reshape((-1,2))
    sigma_em = idata.posterior["sigma_em"].values.reshape((-1,1))

    fig3, ax3 = plt.subplots()
    sigma_plot = np.linspace(0.0,10.0,200)
    ax3.hist(sigma_em, bins = 49, density = True, color="tomato",edgecolor="black")
    ax3.plot(sigma_plot, halfnorm.pdf(sigma_plot), lw = 2, color = "blue") # Add prior plot
    ax3.plot([sigma_em_MLE, sigma_em_MLE], [0, ax3.get_ylim()[1]], "--k", linewidth=2)
    ax3.annotate("MLE solution", xy = (sigma_em_MLE, 0.5*ax3.get_ylim()[1]), xytext = (sigma_em_MLE-2.0, 0.75*ax3.get_ylim()[1]), arrowprops = dict(facecolor="black",shrink=0.05, width = 1.0))
    ax3.set_title("Emulator standard deviation")
    ax3.set_xlabel("Sigma_em")
    ax3.set_ylabel("Density")

    fig4, axs4 = plt.subplots(1,2,figsize=[10,6])
    ls_plot = np.linspace(0.0,4.0,100)
    for i, ax4 in enumerate(axs4):
        ax4.hist(ls[:,i], bins = 49, density = True, color = "tomato", edgecolor="black")
        ax4.plot([ls_MLE[i], ls_MLE[i]],  [0, 0.9*ax4.get_ylim()[1]], "--k", linewidth=2)
        ax4.annotate("MLE solution", xy = (ls_MLE[i], 0.5*ax4.get_ylim()[1]), xytext = (0.1*ax4.get_xlim()[1], 0.95*ax4.get_ylim()[1]), arrowprops = dict(facecolor="black",shrink=0.05, width = 1.0))
        # Note: I think the scipy documentation is incorrect about the inverse gamma pdf
        # I've compared with my own implementation with the pymc docs and it matches, 
        # so I'm confident the below correctly matches the prior
        ax4.plot(ls_plot, invgamma.pdf(ls_plot, 4.0, loc = 0.0, scale = 4.0), lw = 2, color = "blue")
        ax4.set_xlabel("ls_" + str(i))
        ax4.set_ylabel("Density")
  
    N_post = sigma_em.shape[0] # Number of posterior samples
    
#------------------------------------------------------------------------------
#                 Make predictions using the fitted emulator
#------------------------------------------------------------------------------

    # Take a sub-sample of the posterior parameters (Note: If the number of 
    # predictions is greater than the numver of posterior samples set replace = True)
    N_sub_sam = 50
    pred_ind = np.random.choice(N_post,N_sub_sam, replace = False)
    mu_pred = np.empty((N_pred, N_sub_sam),float)
    sigma_pred = np.empty((N_pred, N_sub_sam),float)
    if gp_sample_pred:
        f_pred = np.empty((N_pred, N_sub_sam),float)

    # Take N_sam_pred random samples from the posterior and generate a sample prediction
    for i, post_i in enumerate(pred_ind):
        point = {"sigma_em" : float(sigma_em[post_i,0]), "ls" : ls[post_i,:]}
        if gp_sample_pred: 
            mu_pred[:,i], cov = gp.predict(x_pred_trans, point = point, model = emulator_model, diag = False)
            f_pred[:,i] = np.random.multivariate_normal(mu_pred[:,i], cov)
            sigma_pred[:,i] = np.diagonal(cov)
        else:
            mu_pred[:,i], sigma_pred[:,i] = gp.predict(x_pred_trans.values, point = point, model = emulator_model, diag = True)
    
    sigma_pred = np.sqrt(sigma_pred)
    mu_pred = np.mean(mu_pred,axis=1)
    sigma_pred = np.mean(sigma_pred,axis=1)

#------------------------------------------------------------------------------
#           Transform back onto correct scale then plot predictions
#------------------------------------------------------------------------------

    # Take average across posterior samples to integrate out uncertainty 
    # Convert the mean and standard deviation back onto the true scale
    mu_pred = rescale_output(mu_pred, mu_y = y_mu, sigma_y = y_sd)
    sigma_pred = rescale_output(sigma_pred, sigma_y = y_sd, std = True)

    fig5, ax5 = plt.subplots(1,2,subplot_kw = {"projection" : "3d"})
    ax5[0].plot_surface(x_grid, t_grid, mu_pred.reshape(x_grid.shape), rstride=1, cstride=1, cmap='jet', linewidth=0, antialiased=False, alpha = 0.8)
    ax5[0].scatter(x_train["x"], x_train["d"], y_train, s = 25, color = "red", marker = "x")
    ax5[0].set_xlabel("x (m)", labelpad = 14.0)
    ax5[0].set_ylabel("Beam height (m)", labelpad = 14.0)
    ax5[0].set_zlabel("Displacement (m)", labelpad = 14.0)
    ax5[0].set_title("Bayesian Emulator Mean")
    ax5[1].plot_surface(x_grid, t_grid, mu_pred_MLE.reshape(x_grid.shape), rstride=1, cstride=1, cmap='jet', linewidth=0, antialiased=False, alpha = 0.8)
    ax5[1].scatter(x_train["x"], x_train["d"], y_train, s = 25, color = "red", marker = "x")
    ax5[1].set_xlabel("x (m)", labelpad = 14.0)
    ax5[1].set_ylabel("Beam height (m)", labelpad = 14.0)
    ax5[1].set_zlabel("Displacement (m)", labelpad = 14.0)
    ax5[1].set_title("MLE Emulator Mean")

    fig6, ax6 = plt.subplots(1,2,subplot_kw = {"projection": "3d"})
    ax6[0].plot_surface(x_grid, t_grid, sigma_pred.reshape(x_grid.shape), rstride=1, cstride=1, cmap='jet', linewidth=0, antialiased=False)
    ax6[0].scatter(x_train["x"], x_train["d"], np.zeros(x_train.shape[0]), s = 25, color = "red", marker = "x")
    ax6[0].set_xlabel("x (m)", labelpad = 14.0)
    ax6[0].set_ylabel("Beam depth (m)", labelpad = 14.0)
    ax6[0].set_zlabel("Standard deviation (m)", labelpad = 14.0)
    ax6[0].set_title("Bayesian Emulator Standard deviation")
    ax6[0].ticklabel_format(useOffset=False, style = "plain") # Keep in standard format, not scientific
    ax6[1].plot_surface(x_grid, t_grid, sigma_pred_MLE.reshape(x_grid.shape), rstride=1, cstride=1, cmap='jet', linewidth=0, antialiased=False)
    ax6[1].scatter(x_train["x"], x_train["d"], np.zeros(x_train.shape[0]), s = 25, color = "red", marker = "x")
    ax6[1].set_xlabel("x (m)", labelpad = 14.0)
    ax6[1].set_ylabel("Beam depth (m)", labelpad = 14.0)
    ax6[1].set_zlabel("Standard deviation (m)", labelpad = 14.0)
    ax6[1].set_title("MLE Emulator Standard deviation")
    ax6[1].ticklabel_format(useOffset=False, style = "plain") # Keep in standard format, not scientific
    plt.tight_layout()
    plt.show()
